openapps/csensors/csensors.c

```diff
@@ -0,0 +1,398 @@
+   OpenQueueEntry_t*          pkt;
+   owerror_t                  outcome;
+   uint8_t                    id = cb_vars[cb_get];
+
+   // don't run if not synch
+   if (ieee154e_isSynch() == FALSE) return;
+
+   // don't run on dagroot
+   if (idmanager_getIsDAGroot()) {
+      opentimers_stop(csensors_vars.csensors_resource[id].timerId);
+      return;
+   }
+
+   // create a CoAP RD packet
+   pkt = openqueue_getFreePacketBuffer(COMPONENT_CSENSORS);
+   if (pkt==NULL) {
+      openserial_printError(
+         COMPONENT_CSENSORS,
+         ERR_NO_FREE_PACKET_BUFFER,
+         (errorparameter_t)0,
+         (errorparameter_t)0
+      );
+      openqueue_freePacketBuffer(pkt);
+      return;
+   }
+
+   // take ownership over that packet
+   pkt->creator                   = COMPONENT_CSENSORS;
+   pkt->owner                     = COMPONENT_CSENSORS;
+
+   // CoAP payload
+   csensors_fillpayload(pkt,id);
+
+   packetfunctions_reserveHeaderSize(pkt,2);
+   pkt->payload[0]                = (COAP_OPTION_NUM_CONTENTFORMAT - COAP_OPTION_NUM_URIPATH) << 4
+                                       | 1;
+   pkt->payload[1]                = COAP_MEDTYPE_APPOCTETSTREAM;
+
+   // location-path1 option
+   packetfunctions_reserveHeaderSize(pkt,1+csensors_vars.csensors_resource[id].desc.path1len);
+   memcpy(&pkt->payload[1],csensors_vars.csensors_resource[id].desc.path1val,csensors_vars.csensors_resource[id].desc.path1len);
+   pkt->payload[0]                =  sizeof(csensors_path0)-1;
+   // location-path0 option
+   packetfunctions_reserveHeaderSize(pkt,sizeof(csensors_path0));
+   memcpy(&pkt->payload[1],csensors_path0,sizeof(csensors_path0)-1);
+   pkt->payload[0]                = ((COAP_OPTION_NUM_URIPATH) << 4) | (sizeof(csensors_path0)-1);
+
+   // metadata
+   pkt->l4_destination_port       = WKP_UDP_COAP;
+   pkt->l3_destinationAdd.type    = ADDR_128B;
+   memcpy(&pkt->l3_destinationAdd.addr_128b[0],&ipAddr_ringmaster,16);
+
+   // send
+   outcome = opencoap_send(
+      pkt,
+      COAP_TYPE_NON,
+      COAP_CODE_REQ_PUT,
+      2,
+      &csensors_vars.csensors_resource[id].desc
+   );
+
+   // avoid overflowing the queue if fails
+   if (outcome==E_FAIL) {
+      openqueue_freePacketBuffer(pkt);
+   } else {
+      cb_get = (cb_get+1)%CSENSORSTASKLIST;
+   }
+
+   return;
+}
+
+/**
+\brief Called when receiving a CoAP PUT.
+
+\param[in] period       The period used for reporting data.
+\param[in] id           Resource id in sensors array.
+
+*/
+void csensors_changebehavior(uint32_t period,
+      uint8_t id) {
+
+   uint32_t old_period = csensors_vars.csensors_resource[id].period;
+   float denominator = 65535;
+
+   if (period>0) {
+      csensors_vars.csensors_resource[id].period = period;
+      if (csensors_vars.csensors_resource[id].timerId != MAX_NUM_TIMERS) {
+         opentimers_setPeriod(
+            csensors_vars.csensors_resource[id].timerId,
+            TIME_MS,
+            (uint32_t)(period*(openrandom_get16b()/denominator)));
```

Comment:
I don't understand what you are calculating here. This will result in integer division errors.
Change your period manipulation to rely only on integers